plutus-core/docs/equalsCanonical.md

```diff
@@ -0,0 +1,203 @@
+
+## Misc
+
+We may want to have ways of constructing `canonical` values other than via a `a -> Data` proof, since not everything canonical can be represented as `Data`. We can always add more builtins, but that seems like it might be wasteful, not sure.
+
+We will of course need an opposite builtin to `mkCanonical`, but that's easy, on the Plutus side it's
+
+```
+unCanonical : all a. canonical a -> a
+```
+
+and on the Haskell side it's
+
+```haskell
+    toBuiltinMeaning _ver UnCanonical =
+        makeBuiltinMeaning
+            unCanonicalPlc
+            (\_ _ -> ExBudgetLast mempty)
+      where
+        unCanonicalPlc :: Opaque val (Canonical a) -> Opaque val a
+        unCanonicalPlc = coerce
+```
+
+This function is of course erasable too.
+
+What's the kind of `canonical`? It may seem obvious that it's `* -> *`, but this isn't obvious at all, given that built-in types cannot create a `*` on the metatheory side, only a `#` (this distinction doesn't exist on the implementation side). Do we want to allow pointless built-in types like `[Canonical a]` if we're going to maintain the rule that any built-in type constructs a `#`? Or do we simply add `Canonical` to the `Type` AST and not make it into a built-in type polluting Typed Plutus Core with very Cardano-specific details? We'll need to sort that out too. Also the metatheory currently doesn't support builtins taking functions as arguments (implementation supports arbitrary Plutus types, although some are more ergonomic than others).
+
+In the `EqCanonical` instance for `CekValue` we require each built-in type from the universe to implement `Eq`:
+
+```haskell
+instance (GEq uni, Closed uni, uni `Everywhere` Eq) => EqCanonical (CekValue uni fun ann) where
+```
+
+We may want to require ``uni `Everywhere` EqCanonical`` instead, in case some built-in types don't have decidable equality (currently all do).
+
+Instead of having an `EqCanonical` class we may want to have a class for extracting either a SOP constructor (`VConstr` in `CekValue`) or a constant (`VCon` in `CekValue`) from a value, so that we can reuse it for both `equalsCanonical` and `serialiseCanonical` as otherwise we'd have to introduce another `SerialiseCanonical` class or something. Adding a new type class for every operation definable on values in canonical form would be silly. Maybe we could even just expand the today's `HasConstant` class that currently only allows one to extract a constant out of a value? Whatever answer to this question is going to be we should add `equalsCanonical` and `SerialiseCanonical` simultaneously in order to ensure we're not missing a part of the picture.
```

Comment:
> -- evaluator-independent values, as they might appear in the semantics
> There is some evaluator-independent way of looking at a value
> Builtins have the ability to inspect this form

Hm, we could do that I guess. Except it should be

```
class Reflect value uni fun where
   reflect :: value -> EvaluationResult (InspectableValue uni fun)
```

or something with `InspectableValue` currently only containing equivalents of `VConstr` and `VCon`, because as you say we definitely don't want `isFunction`.

> But it's hard to see how to avoid that if we're going down this path.

By only adding new constructors to `InspectableValue` when we're sure it's worth it? Something horrible needs to happen for us to decide that allowing the user to branch on whether the given value is a function or not is OK.